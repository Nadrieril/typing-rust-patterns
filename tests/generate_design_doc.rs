use std::fmt::Write;

use typing_rust_patterns::*;

#[test]
fn generate_design_doc() -> anyhow::Result<()> {
    let options = RuleOptions::DEFAULT;

    let examples = [
        ("[&x]: &[&T]", "`x: &T`"),
        ("[&x]: &[&mut T]", "`x: &mut T`, move error"),
        ("[&&mut x]: &[&mut T]", "`x: T`"),
        ("[&mut x]: &mut [&T]", "`x: &T`"),
        ("[&mut x]: &[&mut T]", "type error"),
        ("&[[x]]: &[&mut [T]]", "`x: &mut T`, borrow error"),
        (
            "&[[&x]]: &[&mut [T]]",
            "`x: T`, borrow error if we don't use simplification rules",
        ),
        (
            "&[[&mut x]]: &[&mut [T]]",
            "`x: T`, borrow error if we don't use simplification rules",
        ),
        ("[&ref mut x]: &mut [T]", "`x: &mut T`, borrow error"),
    ];

    let mut doc = String::new();
    for (example, comment) in examples {
        let trace = trace_solver(example, options)?;
        writeln!(&mut doc, "- `{example}` => {comment}")?;
        writeln!(&mut doc, "```rust")?;
        writeln!(&mut doc, "{trace}```")?;
        writeln!(&mut doc)?;
    }

    insta::with_settings!({
        snapshot_suffix => "",
        info => &(),
        omit_expression => true,
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_snapshot!(doc);
    });
    Ok(())
}
