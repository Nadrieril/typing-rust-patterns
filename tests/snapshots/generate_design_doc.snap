---
source: tests/generate_design_doc.rs
---
- `[&x]: &[&T]` => `x: &T`
```rust
[&x] @ p: &[&T]
// Applying rule `ConstructorRef`
&x @ &(*p).0: &&T
// Applying rule `Deref(EatOuter)`
x @ *&(*p).0: &T
// Applying rule `Binding`
let x: &T = *&(*p).0

// Final bindings (simplified):
let x: &T = (*p).0;
```

- `[&x]: &[&mut T]` => `x: &mut T`, move error
```rust
[&x] @ p: &[&mut T]
// Applying rule `ConstructorRef`
&x @ &(*p).0: &&mut T
// Applying rule `Deref(EatOuter)`
x @ *&(*p).0: &mut T
// Applying rule `Binding`
let x: &mut T = *&(*p).0

// Final bindings (simplified):
let x: &mut T = (*p).0; // Borrow-check error: CantCopyRefMut
```

- `[&&mut x]: &[&mut T]` => `x: T`
```rust
[&&mut x] @ p: &[&mut T]
// Applying rule `ConstructorRef`
&&mut x @ &(*p).0: &&mut T
// Applying rule `Deref(EatOuter)`
&mut x @ *&(*p).0: &mut T
// Applying rule `Deref(EatOuter)`
x @ **&(*p).0: T
// Applying rule `Binding`
let x: T = **&(*p).0

// Final bindings (simplified):
let x: T = *(*p).0;
```

- `[&mut x]: &mut [&T]` => `x: &T`
```rust
[&mut x] @ p: &mut [&T]
// Applying rule `ConstructorRef`
&mut x @ &mut (*p).0: &mut &T
// Applying rule `Deref(EatOuter)`
x @ *&mut (*p).0: &T
// Applying rule `Binding`
let x: &T = *&mut (*p).0

// Final bindings (simplified):
let x: &T = (*p).0;
```

- `[&mut x]: &[&mut T]` => type error
```rust
[&mut x] @ p: &[&mut T]
// Applying rule `ConstructorRef`
&mut x @ &(*p).0: &&mut T
// Type error for `&mut x @ &(*p).0: &&mut T`: MutabilityMismatch
```

- `&[[x]]: &[&mut [T]]` => `x: &mut T`, borrow error
```rust
&[[x]] @ p: &[&mut [T]]
// Applying rule `Deref(EatOuter)`
[[x]] @ *p: [&mut [T]]
// Applying rule `Constructor`
[x] @ (*p).0: &mut [T]
// Applying rule `ConstructorRef`
x @ &mut (*(*p).0).0: &mut T
// Applying rule `Binding`
let x: &mut T = &mut (*(*p).0).0

// Final bindings (simplified):
let x: &mut T = &mut (*(*p).0).0; // Borrow-check error: MutBorrowBehindSharedBorrow
```

- `&[[&x]]: &[&mut [T]]` => `x: T`, borrow error if we don't use simplification rules
```rust
&[[&x]] @ p: &[&mut [T]]
// Applying rule `Deref(EatOuter)`
[[&x]] @ *p: [&mut [T]]
// Applying rule `Constructor`
[&x] @ (*p).0: &mut [T]
// Applying rule `ConstructorRef`
&x @ &mut (*(*p).0).0: &mut T
// Applying rule `DerefMutWithShared(EatOuter)`
x @ *&*&mut (*(*p).0).0: T
// Applying rule `Binding`
let x: T = *&*&mut (*(*p).0).0

// Final bindings (simplified):
let x: T = (*(*p).0).0;
```

- `&[[&mut x]]: &[&mut [T]]` => `x: T`, borrow error if we don't use simplification rules
```rust
&[[&mut x]] @ p: &[&mut [T]]
// Applying rule `Deref(EatOuter)`
[[&mut x]] @ *p: [&mut [T]]
// Applying rule `Constructor`
[&mut x] @ (*p).0: &mut [T]
// Applying rule `ConstructorRef`
&mut x @ &mut (*(*p).0).0: &mut T
// Applying rule `Deref(EatOuter)`
x @ *&mut (*(*p).0).0: T
// Applying rule `Binding`
let x: T = *&mut (*(*p).0).0

// Final bindings (simplified):
let x: T = (*(*p).0).0;
```

- `[&ref mut x]: &mut [T]` => `x: &mut T`, borrow error
```rust
[&ref mut x] @ p: &mut [T]
// Applying rule `ConstructorRef`
&ref mut x @ &mut (*p).0: &mut T
// Applying rule `DerefMutWithShared(EatOuter)`
ref mut x @ *&*&mut (*p).0: T
// Applying rule `BindingBorrow`
x @ &mut *&*&mut (*p).0: &mut T
// Applying rule `Binding`
let x: &mut T = &mut *&*&mut (*p).0

// Final bindings (simplified):
let x: &mut T = &mut *&(*p).0; // Borrow-check error: MutBorrowBehindSharedBorrow
```
